Add basic driver for DRV8837

kevinhikaruevans opened the pull request:
For issue #296 

Had a couple of questions/problems:

- I removed the blocking parameter from both this class and from the interface, since you (@WattsUp) mentioned that it would not be needed.
- I can't set nSleep low. The orange wire is giving me a logic high.
- After setting the pwm, I'm sleeping the board for 10 microseconds, just to ensure it's being asserted. I'm not quite sure if this is actually needed though.

Commit:
Add basic driver for DRV8837; modify hbridge interface.

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -23,17 +31,60 @@ DRV8837::~DRV8837() {}
+  if (value == 0) {
+    return this->stop(false);
+  } else {
+    if (value < 0) {
```
0.0
Can you write the forward case first? Easier to read

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -23,17 +31,60 @@ DRV8837::~DRV8837() {}
+  } else {
+    if (value < 0) {
+      this->rev = -value;
+      this->fwd = 0.0;
```
There is some implicit casting here. Make it explicit

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -23,17 +31,60 @@ DRV8837::~DRV8837() {}
+    }
+  }
+
+  wait_us(DRV8837_SET_WAIT_TIME_US);
```
Don't need

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -23,17 +31,60 @@ DRV8837::~DRV8837() {}
  */
-mbed_error_status_t DRV8837::stop(bool brake, bool blocking) {
-  return MBED_ERROR_UNSUPPORTED;
+mbed_error_status_t DRV8837::stop(bool brake) {
```
Throw in bool sleep here

kevinhikaruevans replied:
Not quite sure what you mean here. 

As in, keep the blocking argument and sleep the PS if it's true?

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -23,17 +31,60 @@ DRV8837::~DRV8837() {}
+ * @brief Wakes the Hbridge.
+ * 
+ */
+void DRV8837::wake() {
```
nSleep = active low sleep = 1 to wake. Read 7.3.1. Change this to enable and pass in a bool to enable or not

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -23,17 +31,60 @@ DRV8837::~DRV8837() {}
+ * @brief Force the Hbridge to sleep.
+ * 
+ */
+void DRV8837::sleep_immediate() {
```
Functionality moving to enable above

WattsUp reviewed (changes requested):
1. Leave the blocking parameter in the interface but ignore it in this implementation. There could be a future IC that uses that parameter. You can comment out parameters
2. That is odd. Give it to me or Hendrik on Sunday to debug
3. No sleep needed

You don't have your formatter setup. Read the readme

kevinhikaruevans commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -23,17 +31,60 @@ DRV8837::~DRV8837() {}
+    return MBED_ERROR_CODE_INVALID_ARGUMENT;
+  }
+
+  this->wake();
```
Where do I find that document? I assume it's the DRV8837 datasheet?

Commit:
Fix formatting; default nSleep = 0, use usec period instead of freq, clean up

Commit:
readd default parameter

kevinhikaruevans commented:
@WattsUp @micah-priddis 

Please review the updated PR when you guys have a chance. Changelog:
- Used usec period instead of freq
- Re-added `blocking` parameter in `HBridge` interface
- Set default nSleep pin state
- Removed redundant floating pt checks
- Used explicit `f` numeric casts (could also do `(float)`?)
- Changed functionality--added `setSleep(bool sleep)`

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -7,51 +7,38 @@
+DRV8837::DRV8837(PinName fwd, PinName rev, PinName nSleep) :
+  fwd(fwd), rev(rev), nSleep(nSleep, 0) {
+  // set the pwm periods
+  this->fwd.period_us(DRV8837_PWM_PERIOD_US);
```
Add the period to the arguments of the constructor. What if we want one hbridge at 1ms and the other at 1ns

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -7,51 +7,38 @@
-  if (value == 0) {
-    return this->stop(false);
+  if (value < 0) {
+    this->rev = -1.0f * value;
```
This would still be a double as a float * double takes the higher precision of a double.

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -7,51 +7,38 @@
-    return MBED_ERROR_CODE_INVALID_ARGUMENT;
-  }
+mbed_error_status_t DRV8837::set(double value, bool blocking) {
+  this->setSleep(false);
```
Don't touch sleep in here. The function just sets the output not the sleep status.
For the unused bool blocking, it could give you an used warning, you could comment it out: bool /* blocking */

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -60,31 +47,22 @@ mbed_error_status_t DRV8837::set(double value) {
-  wait_us(DRV8837_SET_WAIT_TIME_US);
-  this->sleep_immediate();
+mbed_error_status_t DRV8837::stop(bool brake, bool blocking) {
+  this->setSleep(false);
```
Same about sleep and blocking above.

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.h:
```diff
@@ -15,16 +14,15 @@ class DRV8837 : public HBridge {
 
-  void wake();
-  void sleep_immediate();
+  void setSleep(bool sleep);
```
Should this be a public member function? I think so...

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -60,31 +47,22 @@ mbed_error_status_t DRV8837::set(double value) {
- * 
+ * @brief Sets the HBridge operating mode
+ * @param sleep sets the operating mode to sleep if true, and awakens the
+ * HBridge is false.
```
The sleep signal is not to wake up logic to change the PWM then sleep again. The DS has a poor name. It is an enable output signal. When high the output is given by the input: for, rev, brake, coast. When nSleep is low the output is coasting and ignoring any of the PWM input and enters a low power state called sleeping.

WattsUp reviewed (changes requested):
Just small issue with understanding the function of nSleep signal

WattsUp commented:
Lets leave this open since this is a good change. Someone else can take ownership since @kevinhikaruevans is busy at the moment

Shade-MC committed:
Implemented requested changes

WattsUp commented on CISLibrary/drivers/HBridge/DRV8837.cpp:
```diff
@@ -7,15 +7,17 @@
 
 /**
- * @brief Destroy the DRV8837::DRV8837 object
+ * @brief Eradicate the DRV8837::DRV8837 object
```
Why? Okay...

Shade-MC replied:
is funny

warior4356 committed:
Merge branch 'master' into kevans-DRV8837

WattsUp committed:
Fixed compilatino error
